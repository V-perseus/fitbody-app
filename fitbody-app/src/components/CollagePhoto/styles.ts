import { StyleSheet } from 'react-native'
import globals from '../../config/globals'

const styles = StyleSheet.create({
  collageFrame: {
    height: globals.window.width - 48,
  },
  collageFooter: {
    height: 48,
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignItems: 'center',
    justifyContent: 'center',
  },
  collageImages: {
    // flex: 1,
    height: globals.window.width - 48 - 48,
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
  collageImage: {
    borderWidth: 0.2,
    width: '50%',
  },
  collageFooterItem: {
    width: '50%',
    height: 48,
    alignItems: 'center',
    justifyContent: 'center',
  },
  collageFooterText: {
    fontFamily: globals.fonts.secondary.style.fontFamily,
    color: globals.styles.colors.colorWhite,
    lineHeight: 19.5,
    fontSize: 15,
  },
  footerIcon: {
    color: globals.styles.colors.colorWhite,
    fontSize: 30,
    textAlign: 'center',
  },
  beforeImage: {
    height: '100%',
    width: '100%',
  },
})

export default styles
