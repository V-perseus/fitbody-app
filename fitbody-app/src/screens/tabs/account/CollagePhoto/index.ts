import CollagePhoto from './CollagePhoto'

export default CollagePhoto
