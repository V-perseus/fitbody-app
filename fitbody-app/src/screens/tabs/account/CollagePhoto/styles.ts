import { StyleSheet } from 'react-native'
import globals from '../../../../config/globals'

const styles = StyleSheet.create({
  container: {
    flexGrow: 1,
    backgroundColor: globals.styles.colors.colorWhite,
    paddingTop: 8,
  },
  safe: {
    flex: 1,
    ...globals.window,
  },
  icons: {
    ...globals.header.icons,
  },
  navLabel: {
    color: globals.styles.colors.colorBlack,
    fontSize: 16,
    fontFamily: globals.fonts.primary.semiboldStyle.fontFamily,
  },
  photoContainer: {
    flex: 1,
    width: globals.window.width,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    flexWrap: 'wrap',
    paddingHorizontal: 24,
    marginVertical: 20,
  },
  collageFrame: {
    marginLeft: 10,
    marginRight: 10,
    marginBottom: 20,
    height: 300,
  },
  collageFooter: {
    height: 70,
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignItems: 'center',
    justifyContent: 'center',
  },
  collageImages: {
    height: 230,
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
  collageImage: {
    borderWidth: 0.2,
    width: '50%',
  },
  collageFooterItem: {
    width: '33%',
    alignItems: 'center',
    justifyContent: 'center',
  },
  collageFooterText: {
    fontFamily: globals.fonts.secondary.style.fontFamily,
    color: globals.styles.colors.colorWhite,
    fontSize: 24,
  },
  footerIcon: {
    color: globals.styles.colors.colorWhite,
    fontSize: 30,
    textAlign: 'center',
  },
  beforeImage: {
    height: '100%',
    width: '100%',
  },
})

export default styles
