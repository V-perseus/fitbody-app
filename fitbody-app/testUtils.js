import React from 'react'
import { render as rtlRender } from '@testing-library/react-native'
import { configureStore } from '@reduxjs/toolkit'
import { NavigationContainer } from '@react-navigation/native'
import { createStackNavigator } from '@react-navigation/stack'
import { Provider } from 'react-redux'

// Import your own reducer
import { appReducer } from './src/store'
import { ModalStackScreens } from './src/config/routes/loggedInStack'

// custom render method override
function render(ui, { preloadedState, store = configureStore({ reducer: appReducer, preloadedState }), ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react-native'
// override render method
export { render }

// custom navigation mock
const Stack = createStackNavigator()
export const MockedNavigator = ({
  component,
  params = {},
  preloadedState,
  store = configureStore({ reducer: appReducer, preloadedState }),
}) => {
  return (
    <Provider store={store}>
      <NavigationContainer>
        <Stack.Navigator>
          <Stack.Screen name="MockedScreen" component={component} initialParams={params} />
          <Stack.Screen name="Modals" component={ModalStackScreens} />
        </Stack.Navigator>
      </NavigationContainer>
    </Provider>
  )
}
