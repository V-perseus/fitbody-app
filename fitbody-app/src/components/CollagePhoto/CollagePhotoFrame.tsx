import React, { useState, useEffect } from 'react'
import { View, Text, Image, TouchableOpacity, Platform } from 'react-native'
import LinearGradient from 'react-native-linear-gradient'
import ImagePicker from 'react-native-image-crop-picker'

// Assets
import globals from '../../config/globals'
import styles from './styles'

//Components
import FitBody from '../../../assets/images/logo/fitbody.svg'

import { setErrorMessage } from '../../services/error'

interface ICollagePhotoFrameProps {
  beforeImage: string
  afterImage?: string
}
const CollagePhotoFrame: React.FC<ICollagePhotoFrameProps> = ({ beforeImage, afterImage }) => {
  const [images, setImages] = useState<ICollagePhotoFrameProps>({
    afterImage: undefined,
    beforeImage: '',
  })

  useEffect(() => {
    setImages((prevState) => ({
      ...prevState,
      afterImage,
    }))
  }, [afterImage])

  useEffect(() => {
    setImages((prevState) => ({
      ...prevState,
      beforeImage,
    }))
  }, [beforeImage])

  const cropPhoto = async (url: string, key: string) => {
    try {
      if (Platform.OS === 'ios') {
        const image = await ImagePicker.openCropper({
          path: url,
          width: globals.window.height * 0.75,
          height: globals.window.height,
          includeBase64: true,
          cropperToolbarTitle: 'Hey babe! Adjust your photo to be just perfect!',
          writeTempFile: false, // only needed to read from this file and more performant when including base64
          mediaType: 'photo',
        })
        setImages((prevState) => ({
          ...prevState,
          [key]: image.path,
        }))
      }
    } catch (error: any) {
      console.log('PICKER ERROR', error)
      if (!error.message.includes('User cancelled')) {
        setErrorMessage({ error: error.message || 'Something went wrong.' })
      }
    }
  }

  return (
    <View style={styles.collageFrame}>
      <View style={styles.collageImages}>
        <View style={styles.collageImage}>
          <TouchableOpacity onPress={() => cropPhoto(images.beforeImage, 'beforeImage')}>
            <Image style={styles.beforeImage} source={{ uri: images.beforeImage }} testID="before_img" />
          </TouchableOpacity>
        </View>
        <View style={styles.collageImage}>
          <TouchableOpacity onPress={images.afterImage ? () => cropPhoto(images.afterImage!, 'afterImage') : () => {}}>
            <Image style={styles.beforeImage} source={{ uri: images.afterImage }} />
          </TouchableOpacity>
        </View>
      </View>
      <LinearGradient style={styles.collageFooter} colors={[globals.styles.colors.colorLove, globals.styles.colors.colorLove]}>
        <View style={styles.collageFooterItem}>
          <Text style={styles.collageFooterText}>Before</Text>
        </View>
        <View style={styles.collageFooterItem}>
          <Text style={styles.collageFooterText}>After</Text>
        </View>
        <View style={{ position: 'absolute' }}>
          <FitBody color={globals.styles.colors.colorWhite} width={87} />
        </View>
      </LinearGradient>
    </View>
  )
}

export default CollagePhotoFrame
