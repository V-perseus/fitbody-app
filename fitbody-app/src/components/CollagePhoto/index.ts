import CollagePhotoFrame from './CollagePhotoFrame'

export default CollagePhotoFrame
