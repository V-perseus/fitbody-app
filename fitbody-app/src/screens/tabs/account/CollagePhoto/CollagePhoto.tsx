import React, { useEffect, useRef, useState } from 'react'
import { useSelector } from 'react-redux'
import { Text, View, StatusBar, FlatList } from 'react-native'
import ViewShot from 'react-native-view-shot'
import LinearGradient from 'react-native-linear-gradient'
import Share from 'react-native-share'
import RNFS from 'react-native-fs'
import * as _ from 'lodash'

// styles
import globals from '../../../../config/globals'
import styles from './styles'
import ShareIcon from '../../../../../assets/images/svg/icon/24px/share.svg'

// Components
// import ProgressPhotosItem from '../../../../components/ProgressPhotosItem/ProgressPhotosItem'
import ProgressPhotosRow from '../../../../components/ProgressPhotosRow'
import CollagePhotoFrame from '../../../../components/CollagePhoto/CollagePhotoFrame'
import { HeaderButton } from '../../../../components/Buttons/HeaderButton'

import { progressPhotosDaysSelector } from '../../../../data/progress_photos/selectors'

// types
import { IProgressPhoto, ProgressPhotoViewTypes } from '../../../../data/progress_photos/types'
import { CollagePhotoScreenNavigationProps } from '../../../../config/routes/routeTypes'

export type PhotoRowPhotosType = Record<ProgressPhotoViewTypes, IProgressPhoto>
export interface IPhotoRowType {
  day: string
  photos: PhotoRowPhotosType
}

interface ICollagePhotoProps extends CollagePhotoScreenNavigationProps {}
const CollagePhoto: React.FC<ICollagePhotoProps> = ({ navigation, route }) => {
  const [afterImage, setAfterImage] = useState<string | undefined>(undefined)

  const progressDays = useSelector(progressPhotosDaysSelector)

  const viewShot = useRef<ViewShot>(null)

  /**
   * Navigation props
   */
  useEffect(() => {
    const { goBack } = navigation
    navigation.setOptions({
      headerTransparent: false,
      headerStyle: {
        backgroundColor: globals.styles.colors.colorWhite,
        shadowColor: 'transparent',
        elevation: 0, // for Android
        shadowOffset: {
          width: 0,
          height: 0, // for iOS
        },
      },
      headerTitle: () => <Text style={styles.navLabel}>COLLAGE</Text>,
      headerLeft: () => <HeaderButton onPress={() => goBack()} />,
      headerRight: () => (
        <HeaderButton onPress={shareCollage}>
          <ShareIcon color={globals.styles.colors.colorBlack} />
        </HeaderButton>
      ),
    })
  }, [navigation])

  function collagePhoto(picture: IProgressPhoto | undefined) {
    if (picture) {
      setAfterImage(picture.photo_url || undefined)
    }
  }

  function _renderItem({ item }: { item: IPhotoRowType }) {
    return (
      <ProgressPhotosRow
        createdDate={item.day}
        data={item.photos}
        onPressEvent={(photos: IProgressPhoto | undefined) => collagePhoto(photos)}
      />
    )
  }

  function _keyExtractor(item: IPhotoRowType) {
    return item.day
  }

  const shareCollage = () => {
    viewShot.current?.capture?.().then((uri) => {
      RNFS.readFile(uri, 'base64').then((res) => {
        const base64uri = 'data:image/png;base64,' + res
        Share.open({
          title: 'Fitbody Collage',
          message: 'Check out my Fit Body App Progress!',
          type: 'image/png',
          url: base64uri,
        })
          .then((r) => {
            console.log(r)
          })
          .catch((err) => {
            err && console.log(err)
          })
      })
    })
  }

  const extractedRows = Object.keys(progressDays).map((key) => {
    return { day: key, photos: progressDays[key] }
  })
  const rows: IPhotoRowType[] = _.orderBy(extractedRows, (row) => row.day, ['desc'])

  return (
    <View style={styles.container}>
      <StatusBar barStyle="dark-content" />
      <ViewShot ref={viewShot} style={{ marginHorizontal: 24, height: globals.window.width - 48 }}>
        <CollagePhotoFrame beforeImage={route.params?.picture} afterImage={afterImage} />
      </ViewShot>

      <View style={{ flex: 1, marginTop: 24, backgroundColor: globals.styles.colors.colorWhite }}>
        <View style={styles.photoContainer}>
          {rows && rows.length > 0 && <FlatList data={rows} extraData={rows} keyExtractor={_keyExtractor} renderItem={_renderItem} />}
        </View>
      </View>
      <LinearGradient
        pointerEvents="box-none"
        style={{ top: globals.window.width - 48 + 8 + 24, position: 'absolute', height: 113, width: '100%', borderWidth: 0 }}
        colors={['#0000002c', '#00000000']}
      />
    </View>
  )
}

export default CollagePhoto
