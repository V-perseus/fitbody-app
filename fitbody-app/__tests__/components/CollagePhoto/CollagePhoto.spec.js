import React from 'react'
import { act } from '@testing-library/react-native'
import { render } from '../../../testUtils'

import CollagePhoto from '../../../src/components/CollagePhoto'

describe('CollagePhoto', () => {
  it('renders snapshot correctly', () => {
    const tree = render(<CollagePhoto beforeImage={'../../'} afterImage={''} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render ', async () => {
    const { getByText, getByTestId } = render(<CollagePhoto beforeImage={'../../'} afterImage={''} />)
    const elLabel1 = getByText('Before')
    const elLabel2 = getByText('After')
    await act(async () => {
      const beforeImg = getByTestId('before_img')
      expect(elLabel1).not.toBeNull()
      expect(elLabel2).not.toBeNull()
      expect(beforeImg.props.source).toEqual({ uri: '../../' })
    })
  })
})
