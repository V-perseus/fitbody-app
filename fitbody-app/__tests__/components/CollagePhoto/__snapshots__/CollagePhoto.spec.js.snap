// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollagePhoto renders snapshot correctly 1`] = `
<View
  style={
    Object {
      "height": 702,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "flexWrap": "wrap",
        "height": 654,
      }
    }
  >
    <View
      style={
        Object {
          "borderWidth": 0.2,
          "width": "50%",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "../../",
            }
          }
          style={
            Object {
              "height": "100%",
              "width": "100%",
            }
          }
          testID="before_img"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "borderWidth": 0.2,
          "width": "50%",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "height": "100%",
              "width": "100%",
            }
          }
        />
      </View>
    </View>
  </View>
  <BVLinearGradient
    colors={
      Array [
        4294610389,
        4294929551,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={null}
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "height": 48,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "width": "50%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "BebasNeue",
            "fontSize": 15,
            "lineHeight": 19.5,
          }
        }
      >
        Before
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "width": "50%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "BebasNeue",
            "fontSize": 15,
            "lineHeight": 19.5,
          }
        }
      >
        After
      </Text>
    </View>
    <View
      style={
        Object {
          "position": "absolute",
        }
      }
    >
      <fitbody.svg
        color="#ffffff"
        width={87}
      />
    </View>
  </BVLinearGradient>
</View>
`;
